docs/proposals/20191106-machine-load-balancer.md

```diff
@@ -0,0 +1,121 @@
+
+* Some environments need the ability to create different load balancing topologies e.g. 1 LB for control plane traffic, and another for external traffic to the kubernetes API. Or 1 LB per AZ to reduce inter-AZ traffic.
+
+#### Goals
+
+* Provide an extension point(s) to enable integrating external load balancers.
+* Allow the reuse of an LBP across CAPI providers, ex. AWS Elastic Load Balancers (ELB), F5, Netscaler, etc.
+* Represent the status of load balancer in resource status e.g. 4 of 5 members healthy.
+
+#### Non-Goals
+
+* Providing load balancers for non-control plane workloads i.e Service Type Load Balancer
+* Handling other Infrastructure Providers like networking and firewalls, This CAEP may, however, be merged into a larger CAEP when it is ready.
+
+## Summary
+
+The Machine Load Balancer (MLB) provides an abstraction for implementing load balancing across machines while also allowing the resulting  load balancer address to be used by any CRD.
+
+#### Coupling and Dependencies
+
+* The MLB depends on the [template operator](https://docs.google.com/document/d/1dJdnMozsWJ2ccPXWWkZpFVccFFSI7sYaJ6D25sZFlTU/edit#heading=h.vclg2zerysbp) to plumb the load balancer endpoint from the load balancer CRD, to the cluster CRD
+  * The MLB depends on the CAPI Machine CRD'S but it could be extended to other machine like objects as well.
+* CAPI has no dependency on the MLB, it only waits until something sets a controlPlaneEndpoint 
+
+### Workflow
+
+
+![Figure 1](./images/machine-load-balancer/workflow.png)
+
+
+
+**1 - 4:** Controllers watch the resources they responsible for
+
+**5 - 7:** User or system creates all CRD's *concurrently*
+
+**8 - 9:**  Infra provider spins up infrastructure and then waits until the `spec.controlPlaneEndpoint` is set
+
+**10 -11:** The MLB provisions the load balancer and updates `status.endpoint`
+
+**12: ** The template operator picks up that `status.endpoint` has been updated and copies the value to `spec.controlPlaneEndpoint`
+
+**13+** CAPI infra providers pick up that the `controlPlaneEndpoint` has been set and continue provisioning 
+
+##   Types
+
+```go
+type MachineLoadBalancer struct {  
+  ObjectMeta  
+  TypeMeta    
+  Spec MachineLoadBalancerSpec    
+  Status MachineLoadBalancerStatus
+} 
+type MachineLoadBalancerSpec struct {   
+  Port MachineLoadBalancerPort     
+  // MachineSelector provides a mechanism to select multiple machines based on their labels        // allowing for load balancing across clusters and different network topologies.   
+  MachineSelector Selector    
+  // ProviderRef is a load balancer specific CRD used for configuring the underlying load                       // balancer and/or providing access credentials    
+  ProviderRef corev1.ObjectReference
+} 
+
+type MachineLoadBalancerPort struct {    
+  Port int    
```

Comment:
I believe the Kubernetes API conventions [state that integer types should use `int32` or `int64`, not `int`](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types)

> All public integer fields MUST use the Go (u)int32 or Go (u)int64 types, not (u)int (which is ambiguous depending on target platform). Internal types may use (u)int.

Reply:
It's much more important when using protobuf, since it breaks compatibility between 32-bit and 64-bit platforms, but agree that we should prefer int32/64 over int.